### PR TITLE
Save result to firestore

### PR DESCRIPTION
**Pull request by adarsh-jaiswal**
In this PR, we are inserting test result to firestore.
The steps taken to store data : 
1. Insert a record in firestore from the workflow with status as "CREATED".
2. The workflow calls test-runner endpoint. The test-runner changes the status to "RUNNING".
3. The test-runner saves the test result to firestore with status set to "FINISHED".

Sample schema stored in firestore.

```
{
   "jobId":"cd887934-ec17-11ec-8ea0-0242ac120002",
   "runDate":"06-14-2022",
   "commitHash":"21955383d484c2a37f49bbadefcac77f4085f45d",
   "status":{
      "core_2":"RUNNING",
      "core_32":"FINISHED"
   },
   "testRunnerResults":{
      "core_2":{
         "startTime":"",
         "finishTime":"",
         "testStatistics":[
            {
               "jmhVersion":"1.35",
               "benchmark":"benchmark.AddError.benchMarkAddError",
               "mode":"ss",
               "threads":1,
               "forks":5,
               "jvm":"/opt/twitter_mde/package/TwitterJDK8/3f586d7e75fdd179bb40a70b522bdd37620e2a384354ec8abf539fd74e7d4680/Contents/Home/jre/bin/java",
               "jvmArgs":[
                  
               ],
               "jdkVersion":"1.8.0_275",
               "vmName":"TwitterJDK 64-Bit Server VM",
               "vmVersion":"25.275-b01-jvmci-0.30-Tw8r24b1",
               "warmupIterations":1,
               "warmupTime":"single-shot",
               "warmupBatchSize":50000,
               "measurementIterations":1,
               "measurementTime":"single-shot",
               "measurementBatchSize":5000,
               "primaryMetric":{
                  "score":0.3315709108,
                  "scoreError":0.08851958244831233,
                  "scoreConfidence":[
                     0.2430513283516877,
                     0.4200904932483123
                  ],
                  "scorePercentiles":{
                     "0.0":0.306375988,
                     "50.0":0.327302835,
                     "90.0":0.366146108,
                     "95.0":0.366146108,
                     "99.0":0.366146108,
                     "99.9":0.366146108,
                     "99.99":0.366146108,
                     "99.999":0.366146108,
                     "99.9999":0.366146108,
                     "100.0":0.366146108
                  },
                  "scoreUnit":"s/op",
                  "rawData":[
                     [
                        0.340247069
                     ],
                     [
                        0.366146108
                     ],
                     [
                        0.317782554
                     ],
                     [
                        0.327302835
                     ],
                     [
                        0.306375988
                     ]
                  ]
               },
               "secondaryMetrics":{
                  
               }
            }
         ]
      },
      "core_32":{
         "startTime":"",
         "finishTime":"",
         "testStatistics":[
            {
               
            }
         ]
      }
   }
}
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -26,15 +26,19 @@ target/
...
 /.nb-gradle/
-build/
 !**/src/main/**/build/
 !**/src/test/**/build/
 
 ### VS Code ###
 .vscode/
 
 .gradle/
-build/
```
Comment (chriswr95) [inline .gitignore:37]: why is this needed?

File: src/main/java/graphql/testrunner/document/TestStatistics.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    public String vmVersion;
+    public int warmupIterations;
+    public String warmupTime;
+    public int warmupBatchSize;
+    public int measurementIterations;
+    public String measurementTime;
+    public int measurementBatchSize;
+    public PrimaryMetric primaryMetric;
+    public Map<Object, Object> secondaryMetrics;
+
```
Comment (chriswr95) [inline src/main/java/graphql/testrunner/document/TestStatistics.java:26]: nit: whitespace

File: src/main/java/graphql/testrunner/document/PrimaryMetric.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+    public double score;
+    public double scoreError;
+    public List<Double> scoreConfidence;
+    public Map<String, Double> scorePercentiles;
+    public String scoreUnit;
+
+    @JsonIgnore
+    public List<List<Double>> rawData;
+
```
Comment (chriswr95) [inline src/main/java/graphql/testrunner/document/PrimaryMetric.java:18]: nit: whitespace

File: src/main/java/graphql/testrunner/repository/TestResultRepo.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+public class TestResultRepo {
+
+    public static final Logger LOGGER = Logger.getLogger(TestResultRepo.class.getName());
+    private static final String COLLECTION_NAME = "test-runs";
+
+    @Autowired
+    private Firestore firestore;
+
+    public void updateInitialTestResult(UUID jobId, Map<String, Object> updates) {
+        updateDocument(jobId, updates);
```
Comment (chriswr95) [inline src/main/java/graphql/testrunner/repository/TestResultRepo.java:26]: these functions do the same thing, and take the same parameters, do they need to be seperate?

File: src/main/java/graphql/testrunner/repository/TestResultRepo.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.WriteResult;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class TestResultRepo {
```
Comment (chriswr95) [inline src/main/java/graphql/testrunner/repository/TestResultRepo.java:17]: can you think of a better name for this class, `FirestoreWriter` `TestResultSaver` ...?

File: src/main/java/graphql/testrunner/service/TestResultService.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        updates.put(COMMIT_HASH_KEY, job.getCommitHash());
+        updates.put(STATUS_KEY, RUNNING);
+        updates.put(TEST_START_TIME_KEY, serverTimestamp());
+        testResultRepo.updateInitialTestResult(job.getJobId(), updates);
+    }
+
+    public void saveFinalTestResult(Job job) {
+        Map<String, Object> updates = new HashMap<>();
+        updates.put(STATUS_KEY, FINISHED);
+        updates.put(TEST_FINISH_TIME_KEY, serverTimestamp());
```
Comment (chriswr95) [inline src/main/java/graphql/testrunner/service/TestResultService.java:52]: where does the serverTimestamp() function come from?

File: src/main/java/graphql/testrunner/service/TestRunnerService.java [hunk 7]
```diff
@@ -23,11 +25,17 @@ public class TestRunnerService {
     @Autowired
     private CommandExecutorService commandExecutorService;
 
+    @Autowired
+    private TestResultService testResultService;
+
     @Async
     public void runTest(Job job) throws TestRunnerException {
+        testResultService.saveInitialTestResult(job);
```
Comment (chriswr95) [inline src/main/java/graphql/testrunner/service/TestRunnerService.java:33]: minimal updates to existing classes is a sign of well organized code :)

File: src/test/java/graphql/testrunner/service/TestResultServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import org.mockito.InjectMocks;
+
+class TestResultServiceTest {
+
+    @InjectMocks
+    private TestResultService testResultService;
+
+
+    @Test
+    void saveInitialTestResult() {
```
Comment (chriswr95) [inline src/test/java/graphql/testrunner/service/TestResultServiceTest.java:13]: 😤

File: src/test/java/graphql/testrunner/service/TestRunnerServiceTest.java [hunk 9]
```diff
@@ -52,4 +62,16 @@ void runTestForCustomClasses() {
...
         verify(commandExecutorService).executeCommand(eq(command));
     }
+
+    @Test
+    void readResultJson() {
+        List<TestStatistics> testStatistics = new ArrayList<>();
+        try {
+            testStatistics = new ObjectMapper().readValue(new File("result.json"),
+                new TypeReference<List<TestStatistics>>() {});
+        } catch (IOException e) {
```
Comment (chriswr95) [inline src/test/java/graphql/testrunner/service/TestRunnerServiceTest.java:72]: is this actually part of your test? Looks to me like it can be dropped.

**Comment by chriswr95** [comment]
what is the test coverage for your new classes?

File: src/main/java/graphql/testrunner/repository/TestResultRepo.java (hunk 4 above)
Comment (adarsh-jaiswal) [reply src/main/java/graphql/testrunner/repository/TestResultRepo.java]: done

File: src/main/java/graphql/testrunner/service/TestResultService.java (hunk 6 above)
Comment (adarsh-jaiswal) [reply src/main/java/graphql/testrunner/service/TestResultService.java:51]: Firestore library supplies this function to store timestamp in database.

File: .gitignore (hunk 1 above)
Comment (adarsh-jaiswal) [reply .gitignore]: removed.

File: src/test/java/graphql/testrunner/service/TestRunnerServiceTest.java (hunk 9 above)
Comment (adarsh-jaiswal) [reply src/test/java/graphql/testrunner/service/TestRunnerServiceTest.java]: removed.

**Review by chriswr95 — APPROVED** [review]
